src/hash.jl

```diff
@@ -2,103 +2,54 @@
 ## As usual, check out http://www.lysator.liu.se/~nisse/nettle/nettle.html#Hash-functions
 
 import Base: show
-export HashState, HashAlgorithm, HashAlgorithms
+export HashAlgorithm, update!, digest!, hexdigest!, hash!
 
-# All Hash Algorithms derive from this abstract type
-abstract HashAlgorithm
-
-# This is our rather poorly named list of HashAlgorithm types
-HashAlgorithms = DataType[]
-
-
-# This is the user-facing type that is used to actually hash stuff
-type HashState{T<:HashAlgorithm}
-    ctx::Array{UInt8,1}
+immutable NettleHashState
+    hash_type::NettleHashType
+    state::Array{UInt8,1}
 end
 
-# This is a mirror of the nettle-meta.h:nettle_hash struct
-immutable NettleHash
-    name::Ptr{UInt8}
-    context_size::Cuint
-    digest_size::Cuint
-    block_size::Cuint
-    init::Ptr{Void}     # nettle_hash_init_func
-    update::Ptr{Void}   # nettle_hash_update_func
-    digest::Ptr{Void}   # nettle_hash_digest_func
-end
-
-
-# We're going to load in each nettle_hash struct individually, deriving
-# HashAlgorithm types off of the names we find, and calculating the output
-# and context size from the data members in the C structures
-function hash_init()
-    hash_idx = 1
-    while( true )
-    nhptr = unsafe_load(cglobal(("nettle_hashes",nettle),Ptr{Ptr{Void}}),hash_idx)
-    if nhptr == C_NULL
-        break
-    end
-    nh = unsafe_load(convert(Ptr{NettleHash}, nhptr))
-
-    # Otherwise, we continue on to derive the information from this struct
-    name = symbol(uppercase(bytestring(nh.name)))
-
-    # Load in the Ptr{UInt32}'s
-    ctx_size = nh.context_size
-    dgst_size = nh.digest_size
-    block_size = nh.block_size
-
-    # Save the function pointers as well
-    fptr_init = nh.init
-    fptr_update = nh.update
-    fptr_digest = nh.digest
-
-    # First, create the type itself
-    @eval immutable $name <: HashAlgorithm; end
-
-    # Next, record all the important information about this hash algorithm
-    @eval output_size(::Type{$name}) = $(convert(Int,dgst_size))
-    @eval ctx_size(::Type{$name}) = $(convert(Int,ctx_size))
-    @eval hash_type(::Type{$name}) = $nhptr
-
-    # Generate the init, update, and digest functions while we're at it!
-    # Since we have the function pointers from nh, we'll use those
-    @eval function HashState(::Type{$name})
-        ctx = Array(UInt8, ctx_size($name))
-        ccall($fptr_init,Void,(Ptr{Void},),ctx)
-        HashState{$name}(ctx)
-    end
-
-    @eval function update!(state::HashState{$name},data)
-        ccall($fptr_update,Void,(Ptr{Void},Csize_t,Ptr{UInt8}),state.ctx,sizeof(data),pointer(data))
-        state
+# Constructor for NettleHashState
+function HashAlgorithm(name::AbstractString)
```

Comment:
Again, this looks like a constructor for a type called `HashAlgorithm`, but it returns a `NettleHashState`.  Maybe call the type and the function `Hasher` analogous to my suggestion above for `Encrypter` and `Decrypter`?